include/dmic.h

```diff
@@ -0,0 +1,253 @@
+ *     notice, this list of conditions and the following disclaimer in the
+ *     documentation and/or other materials provided with the distribution.
+ *   * Neither the name of the Intel Corporation nor the
+ *     names of its contributors may be used to endorse or promote products
+ *     derived from this software without specific prior written permission.
+ *
+ * THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS "AS IS"
+ * AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE
+ * IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE
+ * ARE DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT OWNER OR CONTRIBUTORS BE
+ * LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR
+ * CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF
+ * SUBSTITUTE GOODS OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS
+ * INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN
+ * CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE)
+ * ARISING IN ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE
+ * POSSIBILITY OF SUCH DAMAGE.
+ *
+ * Author:	Seppo Ingalsuo <seppo.ingalsuo@linux.intel.com>
+ *		Sathish Kuttan <sathish.k.kuttan@.intel.com>
+ *
+ * SPDX-License-Identifier: BSD-3-Clause
+ */
+
+#ifndef __DMIC_H__
+#define __DMIC_H__
+
+#include <device.h>
+
+#ifdef __cplusplus
+extern "C" {
+#endif
+
+/**
+ * DMIC driver states
+ */
+enum dmic_state {
+	DMIC_STATE_UNINIT,	/* Uninitialized */
+	DMIC_STATE_CONFIGURED,	/* Configured */
+	DMIC_STATE_READY,	/* Ready */
+	DMIC_STATE_ACTIVE,	/* Active */
+	DMIC_STATE_PAUSED,	/* Paused */
+};
+
+/**
+ * DMIC driver trigger commands
+ */
+enum dmic_trigger {
+	DMIC_TRIGGER_STOP,	/* stop stream */
+	DMIC_TRIGGER_START,	/* start stream */
+	DMIC_TRIGGER_PAUSE,	/* pause the stream */
+	DMIC_TRIGGER_RELEASE,	/* release paused stream */
+	DMIC_TRIGGER_RESET,	/* reset */
+};
+
+/**
+ * PDM Channels LEFT / RIGHT
+ */
+enum pdm_lr {
+	PDM_CHAN_LEFT,
+	PDM_CHAN_RIGHT,
+};
+
+/**
+ * PDM Input/Output signal configuration
+ */
+struct pdm_io_cfg {
+	/* parameters global to all PDM controllers */
+
+	/* minimum clock frequency supported by the mic */
+	u32_t	min_pdm_clk_freq;
+	/* maximum clock frequency supported by the mic */
+	u32_t	max_pdm_clk_freq;
+	/* minimum duty cycle in % supported by the mic */
+	u8_t	min_pdm_clk_dc;
+	/* maximum duty cycle in % supported by the mic */
+	u8_t	max_pdm_clk_dc;
+
+	/* parameters unique to each PDM controller */
+
+	/* Bit mask to optionally invert PDM clock */
+	u8_t	pdm_clk_pol;
+	/* Bit mask to optionally invert mic data */
+	u8_t	pdm_data_pol;
+	/* Collection of clock skew values for each PDM port */
+	u32_t	pdm_clk_skew;
+};
+
+/**
+ * Buffers to be provided by the application for received data
+ * NOTE: Only Ping-Pong buffers are supported at this time
+ */
+struct stream_buf {
+	void *buffer[2];	/* pointers to the ping pong buffers */
```

Comment:
IMHO we should have buffer request/release approach, as basic ping-ping implies data copying from the statically allocated buffers.

Reply:
I'll add the definition for buffer request/release approach.

Reply:
Added `read` APIs.
Added SLAB in `configure` for the driver to allocate receive buffers from.